compiler/rustc_trait_selection/src/solve/assembly/mod.rs

```diff
@@ -521,49 +521,92 @@ impl<'tcx> EvalCtxt<'_, 'tcx> {
-            // Excluding IATs and type aliases here as they don't have meaningful item bounds.
-            ty::Alias(ty::Projection | ty::Opaque, alias_ty) => alias_ty,
-        };
+        let _ = self.probe(|_| ProbeKind::NormalizedSelfTyAssembly).enter(|ecx| {
+            let mut self_ty = goal.predicate.self_ty();
+
+            // For some deeply nested `<T>::A::B::C::D` rigid associated type,
+            // we should explore the item bounds for all levels, since the
+            // `associated_type_bounds` feature means that a parent associated
+            // type may carry bounds for a nested associated type.
+            loop {
+                let (kind, alias_ty) = match *self_ty.kind() {
+                    ty::Bool
+                    | ty::Char
+                    | ty::Int(_)
+                    | ty::Uint(_)
+                    | ty::Float(_)
+                    | ty::Adt(_, _)
+                    | ty::Foreign(_)
+                    | ty::Str
+                    | ty::Array(_, _)
+                    | ty::Slice(_)
+                    | ty::RawPtr(_)
+                    | ty::Ref(_, _, _)
+                    | ty::FnDef(_, _)
+                    | ty::FnPtr(_)
+                    | ty::Dynamic(..)
+                    | ty::Closure(..)
+                    | ty::Coroutine(..)
+                    | ty::CoroutineWitness(..)
+                    | ty::Never
+                    | ty::Tuple(_)
+                    | ty::Param(_)
+                    | ty::Placeholder(..)
+                    | ty::Infer(ty::IntVar(_) | ty::FloatVar(_))
+                    | ty::Alias(ty::Inherent, _)
+                    | ty::Alias(ty::Weak, _)
+                    | ty::Error(_) => break,
+                    ty::Infer(ty::FreshTy(_) | ty::FreshIntTy(_) | ty::FreshFloatTy(_))
+                    | ty::Bound(..) => bug!("unexpected self type for `{goal:?}`"),
+
+                    // If we hit infer when normalizing the self type of an alias,
+                    // then bail with ambiguity.
+                    ty::Infer(ty::TyVar(_)) => {
+                        if let Ok(result) = ecx
+                            .evaluate_added_goals_and_make_canonical_response(Certainty::AMBIGUOUS)
+                        {
+                            candidates
+                                .push(Candidate { source: CandidateSource::AliasBound, result });
+                        }
+                        break;
+                    }
 
-        for assumption in
-            self.tcx().item_bounds(alias_ty.def_id).instantiate(self.tcx(), alias_ty.args)
-        {
-            match G::consider_alias_bound_candidate(self, goal, assumption) {
-                Ok(result) => {
-                    candidates.push(Candidate { source: CandidateSource::AliasBound, result })
+                    // Excluding IATs and type aliases here as they don't have meaningful item bounds.
```

Comment:
shouldn't IAT and type aliases be literally unreachable here, given that we structurally normalize the self type?

Reply:
Yeah, we should always be able to project those.